File: src/main/java/dev/goerner/geozen/model/MultiPoint.java
package dev.goerner.geozen.model;

import java.util.ArrayList;

/**
 * A {@link MultiPoint} is a {@link Geometry} that represents a collection of {@link Position Positions} in space. It is
 * defined by a list of {@link Position Positions} and a {@link CoordinateReferenceSystem}.
 */
public class MultiPoint extends Geometry {

	private ArrayList<Position> coordinates;

	/**
	 * Creates a new {@link MultiPoint} with an empty list of {@link Position Positions} and the default WGS 84
	 * {@link CoordinateReferenceSystem}.
	 */
	public MultiPoint() {
		this(new ArrayList<>());
	}

	/**
	 * Creates a new {@link MultiPoint} with the given list of {@link Position Positions} and the default WGS 84
	 * {@link CoordinateReferenceSystem}.
	 *
	 * @param coordinates The list of {@link Position Positions} representing the {@link MultiPoint}.
	 */
	public MultiPoint(ArrayList<Position> coordinates) {
		this(coordinates, CoordinateReferenceSystem.WGS_84);
	}

	/**
	 * Creates a new {@link MultiPoint} with the given list of {@link Position Positions} and the given
	 * {@link CoordinateReferenceSystem}.
	 *
	 * @param coordinates               The list of {@link Position Positions} representing the {@link MultiPoint}.
	 * @param coordinateReferenceSystem The {@link CoordinateReferenceSystem} of the {@link MultiPoint}.
	 */
	public MultiPoint(ArrayList<Position> coordinates, CoordinateReferenceSystem coordinateReferenceSystem) {
		super(coordinateReferenceSystem);
		this.coordinates = coordinates;
	}

	public void addPoint(Position point) {
		this.coordinates.add(point);
	}

	public void addPoint(Point point) {
		this.coordinates.add(point.getCoordinates());
	}

	public ArrayList<Position> getCoordinates() {
		return coordinates;
	}

	public void setCoordinates(ArrayList<Position> coordinates) {
		this.coordinates = coordinates;
	}
}

File: src/main/kotlin/dev/goerner/geozen/model/multi_geometry/MultiPoint.kt
package dev.goerner.geozen.model.multi_geometry

import dev.goerner.geozen.calc.ApproximateDistanceCalculator
import dev.goerner.geozen.calc.PreciseDistanceCalculator
import dev.goerner.geozen.model.CoordinateReferenceSystem
import dev.goerner.geozen.model.Geometry
import dev.goerner.geozen.model.Position
import dev.goerner.geozen.model.Reprojector

/**
 * A [MultiPoint] is a [Geometry] that represents a collection of [Positions][Position] in space. It
 * is defined by a list of [Positions][Position] and a [CoordinateReferenceSystem].
 */
data class MultiPoint
@JvmOverloads
constructor(
    val coordinates: List<Position>,
    override val coordinateReferenceSystem: CoordinateReferenceSystem =
        CoordinateReferenceSystem.WGS_84,
) : Geometry(coordinateReferenceSystem) {

    override fun fastDistanceTo(other: Geometry): Double =
        ApproximateDistanceCalculator.calculate(this, other)

    override fun exactDistanceTo(other: Geometry): Double =
        PreciseDistanceCalculator.calculate(this, other)

    override fun reprojectTo(crs: CoordinateReferenceSystem): Geometry =
        if (coordinateReferenceSystem == crs) this
        else
            MultiPoint(
                coordinates.map { Reprojector.reproject(it, coordinateReferenceSystem, crs) },
                crs,
            )
}
